### PR TITLE
remove rimraf dependency

### PR DESCRIPTION
**Pull request by benmccann**
closes https://github.com/mapbox/node-pre-gyp/issues/692
closes https://github.com/mapbox/node-pre-gyp/pull/707

this removes quite a few dependencies, many of which are deprecated: https://npmgraph.js.org/?q=rimraf@5.0.5

### CODE REVIEW

File: lib/clean.js
Comment (pnappa) [inline lib/clean.js:27]: I believe this is not correct - `rmdir` only deletes a directory if it's empty, which may not be the case here, no? https://nodejs.org/api/fs.html#fsrmdirpath-options-callback

I think it should be (also replacing the try/catch & synchronous filesystem call):
```
fs.rm(to_delete, { recursive: true, force: true }, callback);
```

File: lib/util/napi.js
Comment (pnappa) [inline lib/util/napi.js:169]: Again, this may not work as the directory may not be empty.

File: lib/util/napi.js
Comment (pnappa) [inline lib/util/napi.js:176]: Also here.

**Review by pnappa — CHANGES_REQUESTED** [review]
I prefer this PR to mine (https://github.com/mapbox/node-pre-gyp/pull/707), as removing dependencies is better than updating them.

However, I believe some code needs fixing.

**Comment by pnappa** [comment]
Also, for the same reasons as in my PR, this will be a breaking change (hopefully no-one's actually affected by it), as it drops support for old versions of node. See: https://github.com/mapbox/node-pre-gyp/pull/707#issuecomment-1861812258

Minimum required version is now 14.x (not sure which version specifically)

File: lib/clean.js
Comment (benmccann) [reply lib/clean.js]: I think `rmdir` and `rm` are both correct, but that `rmdir` is deprecated, so I'll update them all. thanks!!

**Comment by benmccann** [comment]
> Minimum required version is now 14.x (not sure which version specifically)

14.14.0

File: lib/util/napi.js
Comment (pnappa) [inline lib/util/napi.js:169]: Apologies if you interpreted this way (I only meant to say don't use rmdirSync as opposed to rmSync, not to switch to the async version too), but by switching this to the asynchronous version the behaviour here has slightly changed. Change back to `fs.rmSync`, if it's not too much bother.

File: lib/util/napi.js
Comment (pnappa) [inline lib/util/napi.js:176]: Ditto, this should be `fs.rmSync`, which matches the behaviour from before.

**Comment by cclauss** [comment]
Please add the modifications from #709 so we can see if the tests pass on modern versions of Node.js.
* #709

File: lib/clean.js
Comment (benmccann) [inline lib/clean.js:27]: I'm not sure why https://github.com/mapbox/node-pre-gyp/pull/707 added this line. I removed it to return the code to it's original state. @pnappa please check to make sure that I'm not overlooking something

Comment (ronilan) [reply lib/clean.js:27]: The [code](https://github.com/mapbox/node-pre-gyp/blame/8b99dc59ddaac4ed5158e75e70ed485aa08b07ce/lib/clean.js) prior to #707 actually errored. #707 fixed this.

In #720 (this PR) the `try .. catch` clause does the required callback.

This is 👍 

(note: test coverage needs revisit for `clean` as used in `rebuild`)

**Comment by cclauss** [comment]
@ronilan @pnappa @acalcutt @benmccann Can you please help me by reviewing the open PRs on this repo?

When you think a pull request is useful and is ready to be merged, ___please consider giving it a positive review___.

If you think a pull request needs work or should not be merged, ___please consider giving it a negative review___.

Every ✅ at the top right of this page gives project maintainers confidence that the proposed changes have been read through and deemed both useful and safe to merge into the codebase.  Lots of 👍 and "_what is the ETA?_" comments are easier for maintainers to ignore than ✔️✔️✔️✔️✔️ from several different reviewers.

Similarly for ❌ on the other side.

Anyone can review a pull request on GitHub.  To do so here:
1. Scroll to the top of this page.
2. Click the `Files changed` tab and read through each file carefully looking for potential issues.
3. Click the `Review changes` button.
4. Click `Approve` (or one of the other options) and add comments only if they have not already been stated in the PR.
5. Click `Submit review` so that your ✅  or ❌ will be added to the list.

File: lib/clean.js
Comment (benmccann) [reply lib/clean.js:27]: thanks! I've added back the call to `callback`

**Comment by ronilan** [comment]
> @ronilan @pnappa @acalcutt @benmccann Can you please help me by reviewing the open PRs on this repo?

@cclauss - sorry, I ran out of play time. My GitHub Sponsors page is here: https://github.com/sponsors/ronilan Good luck to all that treat legacy code maintenance a hobby rather than work.

File: lib/clean.js
Comment (ronilan) [reply lib/clean.js:27]: Early return not needed. We return at bottom when no error.

Comment (benmccann) [reply lib/clean.js:27]: `callback` is called with different values in each of the locations

Comment (ronilan) [reply lib/clean.js:27]: The [callback function](https://github.com/mapbox/node-pre-gyp/blob/master/lib/main.js#L81) receives a single argument - the error object. The [pattern](https://github.com/mapbox/node-pre-gyp/blob/master/lib/main.js#L81) used when there is no error is to call without passing a value.

**Review by pnappa — CHANGES_REQUESTED** [review]

**Comment by pnappa** [comment]
> @ronilan @pnappa @acalcutt @benmccann Can you please help me by reviewing the open PRs on this repo?

Hi Christian, I'm also not available to review PRs, I only made my PRs & reviewed the related PRs as a professional courtesy to try and patch vulnerable packages in dependencies. I no longer depend on this package, and as such can't justify helping during work hours (in my spare time I try to avoid computers). 

`node-argon2` was the package that I used that depended on `node-pre-gyp`, but they migrated to a different, more upkept package (`prebuildify`). I've talked about it in this PR before https://github.com/mapbox/node-pre-gyp/issues/705#issuecomment-1965458957 . I think community effort should be spent on consolidating onto a single product solution rather than limping along with finite resources spread too thin.

Good luck, and thanks for all the fish!

File: lib/clean.js
Comment (benmccann) [reply lib/clean.js:27]: Done. I've addressed both comments

My confusion stemmed from the use of the word async, which I usually associate with the version returning a promise rather than the version taking a callback. I think we should be sorted now though

**Review by pnappa — APPROVED** [review]

**Review by Alexey-Lukin — APPROVED** [review]

**Comment by benmccann** [comment]
@cclauss this one has two approvals, so I think it should be good to merge now

**Comment by cclauss** [comment]
I would rather progress on #739 before we make any removals of dependencies.  We have made a ton of changes and our test coverage is imperfect.  Let's give users a chance to kick the tires of our current state before we replace parts.
* #739

**Comment by cclauss** [comment]
If you are interested in making contributions while we await progress on #739,  please consider...

Updating https://github.com/mapbox/node-pre-gyp/blob/master/README.md to reflect current versions of Node.js and our changes to automated testing.

The repo could use the addition of https://pre-commit.com or https://typicode.github.io/husky to do fast linting, type checking, spell checking, etc. at commit time.

**Comment by benmccann** [comment]
> Updating https://github.com/mapbox/node-pre-gyp/blob/master/README.md to reflect current versions of Node.js and our changes to automated testing.

I've updated the supported versions of Node as part of https://github.com/mapbox/node-pre-gyp/pull/833. I'm not sure there are any change to automated testing that feel relevant to me to include in the readme. The readme never discussed automated testing in the past and readmes generally don't discuss CI setup.

> The repo could use the addition of https://pre-commit.com/ or https://typicode.github.io/husky to do fast linting, type checking, spell checking, etc. at commit time.

I would prefer not to add these. These types of checks are better on the CI in my opinion where they can be run asynchronously as opposed to interrupting the developer workflow.

> look thru the https://github.com/dependabot generated PRs that fail and add comments to those PRs about why they fail and what changes we should consider to make them pass

A lot of this effort would be wasted if done at this stage. PRs like this one remove a lot of our transitive dependencies. Rather than figuring out why an upgrade somewhere down the tree won't succeed we should first merge the PRs that remove or change dependencies. Then we can go through and figure out how to upgrade what's left.

It should be noted that the vast majority of both merged and pending dependabot PRs have no impact on users. I've sent a PR to reduce the noise from dependabot: https://github.com/mapbox/node-pre-gyp/pull/834

> Or create a changelog PR that describes the changes made for our upcoming release.

Done! https://github.com/mapbox/node-pre-gyp/pull/833

> We have made a ton of changes. Let's give users a chance to kick the tires of our current state before we replace parts.

I'm not sure I really agree with this. As you can see in the changelog PR (https://github.com/mapbox/node-pre-gyp/pull/833), we've only updated supported node versions, updated a few dependencies, and made one bug fix. This PR is less than a dozen lines and just changes the method we call to delete a file. Including it in an initial release would simplify the changelog for users because rather than there being two entries stating "updated rimraf" and "removed rimraf" in two different releases we could condense it down simply to "removed rimraf".

**Comment by cclauss** [comment]
`breaking change` worries me on a codebase that is not being used... https://github.com/mapbox/node-pre-gyp/pull/720#issuecomment-2195801250  My years of effort on `node-gyp` encourages me to make small moves.

**Comment by benmccann** [comment]
The breaking change was introducing the `engines` field that requires Node 18. That's already been done in an earlier PR. The APIs used in this PR only require Node 14.14.0 and so don't introduce any new breaking changes.

**Comment by pnappa** [comment]
> The breaking change was introducing the `engines` field that requires Node 18. That's already been done in an earlier PR. The APIs used in this PR only require Node 14.14.0 and so don't introduce any new breaking changes.

That's right. It's a matter of philosophy whether bumping the minimum required node version is a breaking change or not - I think it is, as a user upgrading a package following semver schemes may have their app fail to compile as a result. However, we don't care about them in this scenario, as they're using EOL'd versions of NodeJS, and have much bigger problems to consider (unpatched security vulnerabilities et al).

Others disagree whether it's a breaking change or not - here, no functional run-time behaviour has changed, and as such should be fine to merge.

**Comment by benmccann** [comment]
I agree that bumping the minimum node version is a breaking change, but my point is that's not being done here because we've earlier specified Node 18 as the minimum.

**Review by cclauss — APPROVED** [review]
Thanks for making it clear.  Let's go...

**Comment by 43081j** [comment]
@cclauss any chance we can get a release of this? 👀

**Comment by benmccann** [comment]
The latest release was 1.1.0-dev.1, but this PR was not included in that release. Unfortunately releasing this PR by itself won't help a ton as minizlib also pulls in rimraf. That PR will need to be merged as well to see the benefit.

It's probably worth trying to get some other PRs in as well before doing a new release. @cclauss is there anything I can do to help on that front? I think several PRs in the queue are ready to go

**Pull request merged**
